Compiler crash with custom `unapply` method

## Compiler version

3.6.2, 3.6.3 NEXT

## Minimized code

```scala
case class Data[A]()

trait TCl[A, B]

object Matches:
  def unapply[A](adt: Data[?])[B](using
      ft: TCl[A, B]
  ): Option[Data[A]] = None

given TCl[Unit, String] = new TCl {}

Data() match
  case Matches[Unit](x) => println(x)
```

scastie: https://scastie.scala-lang.org/road21/zxQttu6JT4qdQWgloptmGA/7

## Output (click arrow to expand)
<details>
java.lang.AssertionError: assertion failed
	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:11)
	at dotty.tools.dotc.core.Decorators$.assertingErrorsReported(Decorators.scala:299)
	at dotty.tools.dotc.typer.Applications.loop$1(Applications.scala:1587)
	at dotty.tools.dotc.typer.Applications.loop$1(Applications.scala:1586)
	at dotty.tools.dotc.typer.Applications.unapplyImplicits$1(Applications.scala:1589)
	at dotty.tools.dotc.typer.Applications.typedUnApply(Applications.scala:1638)
	at dotty.tools.dotc.typer.Applications.typedUnApply$(Applications.scala:434)
	at dotty.tools.dotc.typer.Typer.typedUnApply(Typer.scala:145)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3496)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3581)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3658)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3662)
	at dotty.tools.dotc.typer.Typer.typedPattern(Typer.scala:3793)
	at dotty.tools.dotc.typer.Typer.typedCase(Typer.scala:2228)
	at dotty.tools.dotc.typer.Typer.typedCases$$anonfun$1(Typer.scala:2154)
	at dotty.tools.dotc.core.Decorators$.loop$1(Decorators.scala:99)
	at dotty.tools.dotc.core.Decorators$.mapconserve(Decorators.scala:115)
	at dotty.tools.dotc.typer.Typer.typedCases(Typer.scala:2153)
	at dotty.tools.dotc.typer.Typer.$anonfun$39(Typer.scala:2142)
	at dotty.tools.dotc.typer.Applications.harmonic(Applications.scala:2603)
	at dotty.tools.dotc.typer.Applications.harmonic$(Applications.scala:434)
	at dotty.tools.dotc.typer.Typer.harmonic(Typer.scala:145)
	at dotty.tools.dotc.typer.Typer.typedMatchFinish(Typer.scala:2142)
	at dotty.tools.dotc.typer.Typer.typedMatch(Typer.scala:2069)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3511)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3581)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3658)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3662)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3773)
	at dotty.tools.dotc.typer.Namer.typedAheadExpr$$anonfun$1(Namer.scala:1765)
	at dotty.tools.dotc.typer.Namer.typedAhead(Namer.scala:1755)
	at dotty.tools.dotc.typer.Namer.typedAheadExpr(Namer.scala:1765)
	at dotty.tools.dotc.typer.Namer.typedAheadRhs$1$$anonfun$1(Namer.scala:2123)
	at dotty.tools.dotc.inlines.PrepareInlineable$.dropInlineIfError(PrepareInlineable.scala:256)
	at dotty.tools.dotc.typer.Namer.typedAheadRhs$1(Namer.scala:2123)
	at dotty.tools.dotc.typer.Namer.rhsType$1(Namer.scala:2131)
	at dotty.tools.dotc.typer.Namer.cookedRhsType$1(Namer.scala:2150)
	at dotty.tools.dotc.typer.Namer.lhsType$1(Namer.scala:2151)
	at dotty.tools.dotc.typer.Namer.inferredResultType(Namer.scala:2162)
	at dotty.tools.dotc.typer.Namer.inferredType$1(Namer.scala:1808)
	at dotty.tools.dotc.typer.Namer.valOrDefDefSig(Namer.scala:1814)
	at dotty.tools.dotc.typer.Namer$Completer.typeSig(Namer.scala:823)
	at dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:990)
	at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:859)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:175)
	at dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:190)
	at dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:192)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:393)
	at dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:3443)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3468)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3580)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3658)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3662)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3684)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3730)
	at dotty.tools.dotc.typer.Typer.typedBlockStats(Typer.scala:1427)
	at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1431)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3504)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3581)
	at dotty.tools.dotc.typer.ProtoTypes$FunProto.$anonfun$7(ProtoTypes.scala:534)
	at dotty.tools.dotc.typer.ProtoTypes$FunProto.cacheTypedArg(ProtoTypes.scala:457)
	at dotty.tools.dotc.typer.ProtoTypes$FunProto.typedArg(ProtoTypes.scala:535)
	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:1007)
	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:1007)
	at dotty.tools.dotc.typer.Applications$Application.addTyped$1(Applications.scala:688)
	at dotty.tools.dotc.typer.Applications$Application.matchArgs(Applications.scala:756)
	at dotty.tools.dotc.typer.Applications$Application.init(Applications.scala:574)
	at dotty.tools.dotc.typer.Applications$TypedApply.&lt;init&gt;(Applications.scala:882)
	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.&lt;init&gt;(Applications.scala:1006)
	at dotty.tools.dotc.typer.Applications.ApplyTo(Applications.scala:1270)
	at dotty.tools.dotc.typer.Applications.ApplyTo$(Applications.scala:434)
	at dotty.tools.dotc.typer.Typer.ApplyTo(Typer.scala:145)
	at dotty.tools.dotc.typer.Applications.simpleApply$1(Applications.scala:1079)
	at dotty.tools.dotc.typer.Applications.realApply$1$$anonfun$2(Applications.scala:1189)
	at dotty.tools.dotc.typer.Typer$.tryEither(Typer.scala:118)
	at dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:1204)
	at dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:1244)
	at dotty.tools.dotc.typer.Applications.typedApply$(Applications.scala:434)
	at dotty.tools.dotc.typer.Typer.typedApply(Typer.scala:145)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3496)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3581)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3658)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3662)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3711)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3730)
	at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:3160)
	at dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$1(Typer.scala:3484)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3488)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3580)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3658)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3662)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3684)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3730)
	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:3293)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3530)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3581)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3658)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3662)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3773)
	at dotty.tools.dotc.typer.TyperPhase.typeCheck$$anonfun$1(TyperPhase.scala:47)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at dotty.tools.dotc.core.Phases$Phase.monitor(Phases.scala:507)
	at dotty.tools.dotc.typer.TyperPhase.typeCheck(TyperPhase.scala:53)
	at dotty.tools.dotc.typer.TyperPhase.$anonfun$4(TyperPhase.scala:99)
	at scala.collection.Iterator$$anon$6.hasNext(Iterator.scala:479)
	at scala.collection.Iterator$$anon$9.hasNext(Iterator.scala:583)
	at scala.collection.immutable.List.prependedAll(List.scala:152)
	at scala.collection.immutable.List$.from(List.scala:685)
	at scala.collection.immutable.List$.from(List.scala:682)
	at scala.collection.IterableOps$WithFilter.map(Iterable.scala:900)
	at dotty.tools.dotc.typer.TyperPhase.runOn(TyperPhase.scala:98)
	at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:343)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
	at dotty.tools.dotc.Run.runPhases$1(Run.scala:336)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:384)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:396)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:69)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:396)
	at dotty.tools.dotc.Run.compileSources(Run.scala:282)
	at dotty.tools.dotc.Run.compile(Run.scala:267)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:37)
	at dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:141)
	at dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
	at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:196)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:252)
	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:186)
	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:166)
	at sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:241)
	at sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:166)
	at sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:214)
	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:542)
	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:542)
	at sbt.internal.inc.Incremental$.$anonfun$apply$3(Incremental.scala:182)
	at sbt.internal.inc.Incremental$.$anonfun$apply$3$adapted(Incremental.scala:180)
	at sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:458)
	at sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:117)
	at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
	at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
	at sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263)
	at sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:413)
	at sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:500)
	at sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:400)
	at sbt.internal.inc.Incremental$.apply(Incremental.scala:208)
	at sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:542)
	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:496)
	at sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
	at sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:433)
	at sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
	at sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2443)
	at sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2393)
	at sbt.internal.server.BspCompileTask$.$anonfun$compute$1(BspCompileTask.scala:41)
	at sbt.internal.io.Retry$.apply(Retry.scala:47)
	at sbt.internal.io.Retry$.apply(Retry.scala:29)
	at sbt.internal.io.Retry$.apply(Retry.scala:24)
	at sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:41)
	at sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2391)
	at scala.Function1.$anonfun$compose$1(Function1.scala:49)
	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:63)
	at sbt.std.Transform$$anon$4.work(Transform.scala:69)
	at sbt.Execute.$anonfun$submit$2(Execute.scala:283)
	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:24)
	at sbt.Execute.work(Execute.scala:292)
	at sbt.Execute.$anonfun$submit$1(Execute.scala:283)
	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:65)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)```
```
</details>

--- a/compiler/src/dotty/tools/dotc/reporting/messages.scala
+++ b/compiler/src/dotty/tools/dotc/reporting/messages.scala
@@ -2421,12 +2421,14 @@ class ClassCannotExtendEnum(cls: Symbol, parent: Symbol)(using Context) extends 
 }
 
 class NotAnExtractor(tree: untpd.Tree)(using Context) extends PatternMatchMsg(NotAnExtractorID) {
-  def msg(using Context) = i"$tree cannot be used as an extractor in a pattern because it lacks an unapply or unapplySeq method"
+  def msg(using Context) = i"$tree cannot be used as an extractor in a pattern because it lacks an ${hl("unapply")} or ${hl("unapplySeq")} method with the appropriate signature"
   def explain(using Context) =
-    i"""|An ${hl("unapply")} method should be defined in an ${hl("object")} as follow:
+    i"""|An ${hl("unapply")} method should be in an ${hl("object")}, take a single explicit term parameter, and:
         |  - If it is just a test, return a ${hl("Boolean")}. For example ${hl("case even()")}
         |  - If it returns a single sub-value of type T, return an ${hl("Option[T]")}
         |  - If it returns several sub-values T1,...,Tn, group them in an optional tuple ${hl("Option[(T1,...,Tn)]")}
+        |
+        |Additionaly, ${hl("unapply")} or ${hl("unapplySeq")} methods cannot take type parameters after their explicit term parameter.
         |
         |Sometimes, the number of sub-values isn't fixed and we would like to return a sequence.
         |For this reason, you can also define patterns through ${hl("unapplySeq")} which returns ${hl("Option[Seq[T]]")}.

--- a/compiler/src/dotty/tools/dotc/typer/Applications.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Applications.scala
@@ -36,6 +36,7 @@ import annotation.threadUnsafe
 
 import scala.util.control.NonFatal
 import dotty.tools.dotc.inlines.Inlines
+import scala.annotation.tailrec
 
 object Applications {
   import tpd.*
@@ -1525,6 +1526,20 @@ trait Applications extends Compatibility {
     def trySelectUnapply(qual: untpd.Tree)(fallBack: (Tree, TyperState) => Tree): Tree = {
       // try first for non-overloaded, then for overloaded occurrences
       def tryWithName(name: TermName)(fallBack: (Tree, TyperState) => Tree)(using Context): Tree =
+        /** Returns `true` if there are type parameters after the last explicit
+         *  (non-implicit) term parameters list.
+         */
+        @tailrec
+        def hasTrailingTypeParams(paramss: List[List[Symbol]], acc: Boolean = false): Boolean =
+          paramss match
+            case Nil => acc
+            case params :: rest =>
+              val newAcc =
+                params match
+                  case param :: _ if param.isType => true
+                  case param :: _ if param.isTerm && !param.isOneOf(GivenOrImplicit) => false
+                  case _ => acc
+              hasTrailingTypeParams(paramss.tail, newAcc)
 
         def tryWithProto(qual: untpd.Tree, targs: List[Tree], pt: Type)(using Context) =
           val proto = UnapplyFunProto(pt, this)
@@ -1532,7 +1547,13 @@ trait Applications extends Compatibility {
           val result =
             if targs.isEmpty then typedExpr(unapp, proto)
             else typedExpr(unapp, PolyProto(targs, proto)).appliedToTypeTrees(targs)
-          if !result.symbol.exists
+          if result.symbol.exists && hasTrailingTypeParams(result.symbol.paramSymss) then
+            // We don't accept `unapply` or `unapplySeq` methods with type
+            // parameters after the last explicit term parameter because we
+            // can't encode them: `UnApply` nodes cannot take type paremeters.
+            // See #22550 and associated test cases.
+            notAnExtractor(result)
+          else if !result.symbol.exists
              || result.symbol.name == name
              || ctx.reporter.hasErrors
           then result

--- a/tests/neg/22550.check
+++ b/tests/neg/22550.check
@@ -1,0 +1,37 @@
+-- [E127] Pattern Match Error: tests/neg/22550.scala:8:9 ---------------------------------------------------------------
+8 |    case Matches(x) => println(x) // error // error
+  |         ^^^^^^^
+  |Matches cannot be used as an extractor in a pattern because it lacks an unapply or unapplySeq method with the appropriate signature
+  |---------------------------------------------------------------------------------------------------------------------
+  | Explanation (enabled by `-explain`)
+  |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+  | An unapply method should be in an object, take a single explicit term parameter, and:
+  |   - If it is just a test, return a Boolean. For example case even()
+  |   - If it returns a single sub-value of type T, return an Option[T]
+  |   - If it returns several sub-values T1,...,Tn, group them in an optional tuple Option[(T1,...,Tn)]
+  |
+  | Additionaly, unapply or unapplySeq methods cannot take type parameters after their explicit term parameter.
+  |
+  | Sometimes, the number of sub-values isn't fixed and we would like to return a sequence.
+  | For this reason, you can also define patterns through unapplySeq which returns Option[Seq[T]].
+  | This mechanism is used for instance in pattern case List(x1, ..., xn)
+   ---------------------------------------------------------------------------------------------------------------------
+-- [E006] Not Found Error: tests/neg/22550.scala:8:31 ------------------------------------------------------------------
+8 |    case Matches(x) => println(x) // error // error
+  |                               ^
+  |                               Not found: x
+  |---------------------------------------------------------------------------------------------------------------------
+  | Explanation (enabled by `-explain`)
+  |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+  | Each identifier in Scala needs a matching declaration. There are two kinds of
+  | identifiers: type identifiers and value identifiers. Value identifiers are introduced
+  | by `val`, `def`, or `object` declarations. Type identifiers are introduced by `type`,
+  | `class`, `enum`, or `trait` declarations.
+  |
+  | Identifiers refer to matching declarations in their environment, or they can be
+  | imported from elsewhere.
+  |
+  | Possible reasons why no matching declaration was found:
+  |  - The declaration or the use is mis-spelt.
+  |  - An import is missing.
+   ---------------------------------------------------------------------------------------------------------------------

--- a/tests/neg/22550.scala
+++ b/tests/neg/22550.scala
@@ -1,0 +1,8 @@
+//> using options -explain
+
+object Matches:
+  def unapply(y: Any)[T]: Option[Any] = None
+
+def main =
+  42 match
+    case Matches(x) => println(x) // error // error

--- a/tests/neg/22550b.check
+++ b/tests/neg/22550b.check
@@ -1,0 +1,37 @@
+-- [E127] Pattern Match Error: tests/neg/22550b.scala:16:9 -------------------------------------------------------------
+16 |    case Matches[Unit](x) => println(x) // error // error
+   |         ^^^^^^^^^^^^
+   |Matches[Unit] cannot be used as an extractor in a pattern because it lacks an unapply or unapplySeq method with the appropriate signature
+   |--------------------------------------------------------------------------------------------------------------------
+   | Explanation (enabled by `-explain`)
+   |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+   | An unapply method should be in an object, take a single explicit term parameter, and:
+   |   - If it is just a test, return a Boolean. For example case even()
+   |   - If it returns a single sub-value of type T, return an Option[T]
+   |   - If it returns several sub-values T1,...,Tn, group them in an optional tuple Option[(T1,...,Tn)]
+   |
+   | Additionaly, unapply or unapplySeq methods cannot take type parameters after their explicit term parameter.
+   |
+   | Sometimes, the number of sub-values isn't fixed and we would like to return a sequence.
+   | For this reason, you can also define patterns through unapplySeq which returns Option[Seq[T]].
+   | This mechanism is used for instance in pattern case List(x1, ..., xn)
+    --------------------------------------------------------------------------------------------------------------------
+-- [E006] Not Found Error: tests/neg/22550b.scala:16:37 ----------------------------------------------------------------
+16 |    case Matches[Unit](x) => println(x) // error // error
+   |                                     ^
+   |                                     Not found: x
+   |--------------------------------------------------------------------------------------------------------------------
+   | Explanation (enabled by `-explain`)
+   |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+   | Each identifier in Scala needs a matching declaration. There are two kinds of
+   | identifiers: type identifiers and value identifiers. Value identifiers are introduced
+   | by `val`, `def`, or `object` declarations. Type identifiers are introduced by `type`,
+   | `class`, `enum`, or `trait` declarations.
+   |
+   | Identifiers refer to matching declarations in their environment, or they can be
+   | imported from elsewhere.
+   |
+   | Possible reasons why no matching declaration was found:
+   |  - The declaration or the use is mis-spelt.
+   |  - An import is missing.
+    --------------------------------------------------------------------------------------------------------------------

--- a/tests/neg/22550b.scala
+++ b/tests/neg/22550b.scala
@@ -1,0 +1,16 @@
+//> using options -explain
+
+case class Data[A]()
+
+trait TCl[A, B]
+
+object Matches:
+  def unapply[A](adt: Data[?])[B](using
+      ft: TCl[A, B]
+  ): Option[Data[A]] = None
+
+given TCl[Unit, String] = new TCl {}
+
+def main =
+  Data() match
+    case Matches[Unit](x) => println(x) // error // error

--- a/tests/neg/bad-unapplies.check
+++ b/tests/neg/bad-unapplies.check
@@ -10,13 +10,13 @@
 -- [E127] Pattern Match Error: tests/neg/bad-unapplies.scala:23:9 ------------------------------------------------------
 23 |    case B("2") => // error (cannot be used as an extractor)
    |         ^
-   |         B cannot be used as an extractor in a pattern because it lacks an unapply or unapplySeq method
+   |B cannot be used as an extractor in a pattern because it lacks an unapply or unapplySeq method with the appropriate signature
    |
    | longer explanation available when compiling with `-explain`
 -- [E127] Pattern Match Error: tests/neg/bad-unapplies.scala:24:9 ------------------------------------------------------
 24 |    case D("2") => // error (cannot be used as an extractor)
    |         ^
-   |         D cannot be used as an extractor in a pattern because it lacks an unapply or unapplySeq method
+   |D cannot be used as an extractor in a pattern because it lacks an unapply or unapplySeq method with the appropriate signature
    |
    | longer explanation available when compiling with `-explain`
 -- [E050] Type Error: tests/neg/bad-unapplies.scala:25:9 ---------------------------------------------------------------

--- a/tests/neg/i18684.check
+++ b/tests/neg/i18684.check
@@ -67,14 +67,16 @@
 -- [E127] Pattern Match Error: tests/neg/i18684.scala:12:6 -------------------------------------------------------------
 12 |  val inner(x) = 3 // error
    |      ^^^^^
-   |      Test.inner cannot be used as an extractor in a pattern because it lacks an unapply or unapplySeq method
+   |Test.inner cannot be used as an extractor in a pattern because it lacks an unapply or unapplySeq method with the appropriate signature
    |--------------------------------------------------------------------------------------------------------------------
    | Explanation (enabled by `-explain`)
    |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
-   | An unapply method should be defined in an object as follow:
+   | An unapply method should be in an object, take a single explicit term parameter, and:
    |   - If it is just a test, return a Boolean. For example case even()
    |   - If it returns a single sub-value of type T, return an Option[T]
    |   - If it returns several sub-values T1,...,Tn, group them in an optional tuple Option[(T1,...,Tn)]
+   |
+   | Additionaly, unapply or unapplySeq methods cannot take type parameters after their explicit term parameter.
    |
    | Sometimes, the number of sub-values isn't fixed and we would like to return a sequence.
    | For this reason, you can also define patterns through unapplySeq which returns Option[Seq[T]].

